File: openHAB/GenericUITableViewCell.m
//
//  GenericUITableViewCell.m
//  openHAB
//
//  Created by Victor Belov on 15/01/14.
//  Copyright (c) 2014 Victor Belov. All rights reserved.
//

#import "GenericUITableViewCell.h"
#import "OpenHABLinkedPage.h"

@interface GenericUITableViewCell() {
    NSDictionary *namedColors;
}
@end

@implementation GenericUITableViewCell
@synthesize widget, textLabel, detailTextLabel, disclosureConstraints;

- (id)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        self.textLabel = (UILabel *)[self viewWithTag:101];
        self.detailTextLabel = (UILabel *)[self viewWithTag:100];
        self.selectionStyle = UITableViewCellSelectionStyleNone;
        self.separatorInset = UIEdgeInsetsZero;
        namedColors = @{
                        @"maroon"     : @"#800000",
                        @"red"        : @"#ff0000",
                        @"orange"     : @"#ffa500",
                        @"olive"      : @"#808000",
                        @"yellow"     : @"#ffff00",
                        @"purple"     : @"#800080",
                        @"fuchsia"    : @"#ff00ff",
                        @"white"      : @"#ffffff",
                        @"lime"       : @"#00ff00",
                        @"green"      : @"#008000",
                        @"navy"       : @"#000080",
                        @"blue"       : @"#0000ff",
                        @"teal"       : @"#008080",
                        @"aqua"       : @"#00ffff",
                        @"black"      : @"#000000",
                        @"silver"     : @"#c0c0c0",
                        @"gray"       : @"#808080"
                        };
    }
    return self;
}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated
{
    [super setSelected:selected animated:animated];

    // Configure the view for the selected state
}

- (void)loadWidget:(OpenHABWidget *)widgetToLoad
{
    self.widget = widgetToLoad;
    if (widget.linkedPage != nil) {
        self.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
        self.selectionStyle = UITableViewCellSelectionStyleBlue;
//        self.userInteractionEnabled = YES;
    } else {
        self.accessoryType = UITableViewCellAccessoryNone;
        self.selectionStyle = UITableViewCellSelectionStyleNone;
//        self.userInteractionEnabled = NO;
    }
        
}

- (void)displayWidget
{
    self.textLabel.text = [self.widget labelText];
    if ([self.widget labelValue] != nil)
        self.detailTextLabel.text = [self.widget labelValue];
    else
        self.detailTextLabel.text = nil;
    [self.detailTextLabel sizeToFit];
    // Clean any detailTextLabel constraints we set before, or they will start to interfere with new ones because of UITableViewCell caching
    if (self.disclosureConstraints != nil) {
        [self removeConstraints:disclosureConstraints];
        disclosureConstraints = nil;
    }
    if (self.accessoryType == UITableViewCellAccessoryNone) {
        // If accessory is disabled, set detailTextLabel (widget value) constraing 20px to the right for padding to the right side of table view
        if (self.detailTextLabel != nil) {
            self.disclosureConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"[detailTextLabel]-20.0-|" options:0 metrics:nil views:NSDictionaryOfVariableBindings(detailTextLabel)];
            [self addConstraints:disclosureConstraints];
        }
    } else {
        // If accessory is enabled, set detailTextLabel (widget value) constraint 0px to the right
        if (self.detailTextLabel != nil) {
            self.disclosureConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"[detailTextLabel]|" options:0 metrics:nil views:NSDictionaryOfVariableBindings(detailTextLabel)];
            [self addConstraints:disclosureConstraints];
        }
    }
    if (self.widget.valuecolor != nil) {
        [self.detailTextLabel setTextColor:[self colorFromHexString:self.widget.valuecolor]];
    } else {
        self.detailTextLabel.textColor = [UIColor lightGrayColor];
    }
    if (self.widget.labelcolor != nil) {
        [self.textLabel setTextColor:[self colorFromHexString:self.widget.labelcolor]];
    } else {
        self.textLabel.textColor = [UIColor blackColor];
    }
}

// This is to fix possible different sizes of user icons - we fix size and position of UITableViewCell icons
- (void)layoutSubviews {
    [super layoutSubviews];
    self.imageView.frame = CGRectMake(13,5,32,32);
}

- (NSString *)namedColorToHexString:(NSString *)namedColor
{
    return namedColors[[namedColor lowercaseString]];
}

- (UIColor *)colorFromHexString:(NSString *)hexString
{
    NSString *colorString = hexString;
    if (![hexString hasPrefix:@"#"]) {
        NSString *namedColor = [self namedColorToHexString:hexString];
        if (namedColor) {
            colorString = namedColor;
        }
    }
    
    unsigned rgbValue = 0;
    NSScanner *scanner = [NSScanner scannerWithString:colorString];
    [scanner setScanLocation:1]; // bypass '#' character
    [scanner scanHexInt:&rgbValue];
    return [UIColor colorWithRed:((rgbValue & 0xFF0000) >> 16)/255.0 green:((rgbValue & 0xFF00) >> 8)/255.0 blue:(rgbValue & 0xFF)/255.0 alpha:1.0];
}

@end

File: openHAB/GenericUITableViewCell.swift
// Copyright (c) 2010-2021 Contributors to the openHAB project
//
// See the NOTICE file(s) distributed with this work for additional
// information.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0
//
// SPDX-License-Identifier: EPL-2.0

import Kingfisher
import OpenHABCore
import UIKit

protocol GenericCellCacheProtocol: UITableViewCell {
    func invalidateCache()
}

class GenericUITableViewCell: UITableViewCell {
    private var _widget: OpenHABWidget!
    var widget: OpenHABWidget! {
        get {
            _widget
        }
        set(widget) {
            _widget = widget

            if _widget.linkedPage != nil {
                accessoryType = .disclosureIndicator
                selectionStyle = .blue
                //        self.userInteractionEnabled = YES;
            } else {
                accessoryType = .none
                selectionStyle = .none
                //        self.userInteractionEnabled = NO;
            }

            customTextLabel?.textColor = !(_widget.labelcolor.isEmpty) ? UIColor(fromString: _widget.labelcolor) : .ohLabel
            customDetailTextLabel?.textColor = !(_widget.valuecolor.isEmpty) ? UIColor(fromString: _widget.valuecolor) : .ohSecondaryLabel
        }
    }

    @IBOutlet private(set) var customTextLabel: UILabel!
    @IBOutlet private(set) var customDetailTextLabel: UILabel!
    @IBOutlet private(set) var customDetailTextLabelConstraint: NSLayoutConstraint!

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        initialize()
    }

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        initialize()
    }

    func initialize() {
        selectionStyle = .none
        separatorInset = .zero
    }

    // This is to fix possible different sizes of user icons - we fix size and position of UITableViewCell icons
    override func layoutSubviews() {
        super.layoutSubviews()
        imageView?.frame = CGRect(x: 13, y: 5, width: 32, height: 32)
    }

    func displayWidget() {
        customTextLabel?.text = widget?.labelText
        customDetailTextLabel?.text = widget?.labelValue ?? ""
        customDetailTextLabel?.sizeToFit()

        if customDetailTextLabel != nil, customDetailTextLabelConstraint != nil {
            if accessoryType == .none {
                // If accessory is disabled, set detailTextLabel (widget value) constraint 20px to the right for padding to the right side of table view
                customDetailTextLabelConstraint.constant = 20.0
            } else {
                // If accessory is enabled, set detailTextLabel (widget value) constraint 5px to the right
                customDetailTextLabelConstraint.constant = 5.0
            }
        }
    }

    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)
    }

    override func prepareForReuse() {
        super.prepareForReuse()
        imageView?.kf.cancelDownloadTask()
        imageView?.image = nil
    }
}
